### PR TITLE
feat: added support for local state schema in document-model editor

### PR DESCRIPTION
**Pull request by gpuente**
## Description:

- added support for local state schema in document-model editor

NOTE: this PR requires a new release from https://github.com/powerhouse-inc/document-model with the latest changes

Screenshot 2023-12-15 at 15 55 27

### Document:

```json
{
  "name": "",
  "documentType": "powerhouse/document-model",
  "revision": 12,
  "created": "2023-12-14T19:02:53.112Z",
  "lastModified": "2023-12-14T19:06:38.109Z",
  "attachments": {},
  "state": {
    "global": {
      "id": "budget",
      "name": "MyDocument",
      "extension": "phxyz",
      "description": "",
      "author": {
        "name": "powerhouse",
        "website": "https://www.powerhouse.inc/"
      },
      "specifications": [
        {
          "version": 1,
          "changeLog": [],
          "state": {
            "global": {
              "schema": "type MyDocumentState {\n\tcolor: String!\n    isMainDoc: Boolean\n}",
              "initialValue": "{\"color\":\"yellow\",\"isMainDoc\":false}",
              "examples": []
            },
            "local": {
              "schema": "type MyDocumentLocalState {\n\ttotal: Int!\n    title: String\n}",
              "initialValue": "{\"total\":123,\"title\":\"Some Title\"}",
              "examples": []
            }
          },
          "modules": []
        }
      ]
    },
    "local": {}
  },
  "initialState": {
    "name": "",
    "documentType": "powerhouse/document-model",
    "revision": 0,
    "created": "2023-12-14T19:02:53.112Z",
    "lastModified": "2023-12-14T19:02:53.112Z",
    "attachments": {},
    "state": {
      "global": {
        "id": "",
        "name": "",
        "extension": "",
        "description": "",
        "author": {
          "name": "",
          "website": ""
        },
        "specifications": [
          {
            "version": 1,
            "changeLog": [],
            "state": {
              "global": {
                "schema": "",
                "initialValue": "",
                "examples": []
              },
              "local": {
                "schema": "",
                "initialValue": "",
                "examples": []
              }
            },
            "modules": []
          }
        ]
      },
      "local": {}
    }
  },
  "operations": {
    "global": [
      {
        "type": "SET_MODEL_ID",
        "input": {
          "id": ""
        },
        "scope": "global",
        "index": 0,
        "timestamp": "2023-12-14T19:02:53.269Z",
        "hash": "Fd20qtObIUDJwJHse6VqFK8ObWY="
      },
      {
        "type": "SET_MODEL_NAME",
        "input": {
          "name": "MyDocument"
        },
        "scope": "global",
        "index": 1,
        "timestamp": "2023-12-14T19:02:58.088Z",
        "hash": "NzxHl/yfrEHYzcCooVCR89UeL7U="
      },
      {
        "type": "SET_MODEL_ID",
        "input": {
          "id": "budget"
        },
        "scope": "global",
        "index": 2,
        "timestamp": "2023-12-14T19:03:06.707Z",
        "hash": "EyZWNrOjL5OaLw1HFv4XGbMWCtc="
      },
      {
        "type": "SET_AUTHOR_NAME",
        "input": {
          "authorName": "powerhouse"
        },
        "scope": "global",
        "index": 3,
        "timestamp": "2023-12-14T19:03:19.226Z",
        "hash": "W0PDOq8YzkKQJBiZil+VA/pDxPk="
      },
      {
        "type": "SET_AUTHOR_WEBSITE",
        "input": {
          "authorWebsite": "https://www.powerhouse.inc/"
        },
        "scope": "global",
        "index": 4,
        "timestamp": "2023-12-14T19:03:29.803Z",
        "hash": "yE0d869PENXbi64jYwJPWvsNAXY="
      },
      {
        "type": "SET_MODEL_EXTENSION",
        "input": {
          "extension": "phxyz"
        },
        "scope": "global",
        "index": 5,
        "timestamp": "2023-12-14T19:03:38.395Z",
        "hash": "gZ2kxPGO4HuVNJeMLipzm24RMoM="
      },
      {
        "type": "SET_STATE_SCHEMA",
        "input": {
          "schema": "type MyDocumentState {\n\tcolor: String!\n    isMainDoc: Boolean\n}",
          "scope": "global"
        },
        "scope": "global",
        "index": 6,
        "timestamp": "2023-12-14T19:04:48.474Z",
        "hash": "ebsk3psAKMZZIQqDEFVf7c417/0="
      },
      {
        "type": "SET_INITIAL_STATE",
        "input": {
          "initialValue": "{\"color\":\"yellow\",\"isMainDoc\":false}",
          "scope": "global"
        },
        "scope": "global",
        "index": 7,
        "timestamp": "2023-12-14T19:05:06.507Z",
        "hash": "x/5xqkE2T2cCzq3CBtfrvght8+M="
      },
      {
        "type": "SET_STATE_SCHEMA",
        "input": {
          "schema": "type MyDocumentLocalState {\n\ttotal: Int!\n    \n}",
          "scope": "local"
        },
        "scope": "global",
        "index": 8,
        "timestamp": "2023-12-14T19:05:59.070Z",
        "hash": "nDoS+G8J+Or/CrGpTXxnWGeAU/M="
      },
      {
        "type": "SET_STATE_SCHEMA",
        "input": {
          "schema": "type MyDocumentLocalState {\n\ttotal: Int!\n    title: String\n}",
          "scope": "local"
        },
        "scope": "global",
        "index": 9,
        "timestamp": "2023-12-14T19:06:11.053Z",
        "hash": "3YUjH48u5XTovZvbmFsIkdkABD8="
      },
      {
        "type": "SET_STATE_SCHEMA",
        "input": {
          "schema": "type MyDocumentLocalState {\n\ttotal: Int!\n    title: String\n}",
          "scope": "local"
        },
        "scope": "global",
        "index": 10,
        "timestamp": "2023-12-14T19:06:21.013Z",
        "hash": "3YUjH48u5XTovZvbmFsIkdkABD8="
      },
      {
        "type": "SET_INITIAL_STATE",
        "input": {
          "initialValue": "{\"total\":123,\"title\":\"Some Title\"}",
          "scope": "local"
        },
        "scope": "global",
        "index": 11,
        "timestamp": "2023-12-14T19:06:38.109Z",
        "hash": "BCjuT5UFMAzHHeT2LRhTzfLvBjI="
      }
    ],
    "local": []
  }
}
```

### CODE REVIEW

File: editors/document-model/editor.tsx [hunk 1]
```diff
@@ -38,8 +37,15 @@ function Editor(props: IProps) {
 
     const { state } = document;
 
+    console.log('document', document);
```
Comment (acaldas) [inline editors/document-model/editor.tsx:40]: Debug console log

File: editors/document-model/useSchemaEditor.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        React.SetStateAction<SchemaState | undefined>
+    >;
+    specification?: DocumentSpecification;
+    setInitialValue: React.Dispatch<React.SetStateAction<JSON>>;
+    initialValue: JSON;
+}
+
+export const useSchemaEditor = (props: UseSchemaEditorProps): SchemaResult => {
+    const { state, scope, setInitialState } = props;
+
```
Comment (acaldas) [inline editors/document-model/useSchemaEditor.ts:37]: I think it's worth it to create a variable for`state[scope]` and `specification?.state[scope]` to make the rest of the code simpler.

**Review by acaldas — COMMENTED** [review]
On setSchemaState we should check if the schema is empty and force it to just be `type ${pascalCase(name)}${scopeStateName}State` without brackets.

File: editors/document-model/editor.tsx (hunk 1 above)
Comment (gpuente) [reply editors/document-model/editor.tsx]: fixed: 13e334dc9004e062e5dd4930d922879a4704ca95

File: editors/document-model/useSchemaEditor.ts (hunk 2 above)
Comment (gpuente) [reply editors/document-model/useSchemaEditor.ts:37]: fixed: 13e334dc9004e062e5dd4930d922879a4704ca95

**Review by acaldas — APPROVED** [review]

**Pull request merged**